File: server/shared/typescript/geo-abstractions/src/interfaces/i-geo-reference.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import type { Country } from "../generated/country.g.js";
import type { Currency } from "../generated/currency.g.js";
import type { GeopoliticalEntity } from "../generated/geopolitical-entity.g.js";
import type { Language } from "../generated/language.g.js";
import type { Locale } from "../generated/locale.g.js";
import type { Subdivision } from "../generated/subdivision.g.js";
import type { Timezone } from "../generated/timezone.g.js";
import type { CountryCode } from "../generated/typed-codes/country-code.g.js";
import type { CurrencyCode } from "../generated/typed-codes/currency-code.g.js";
import type { GeopoliticalEntityCode } from "../generated/typed-codes/geopolitical-entity-code.g.js";
import type { LanguageCode } from "../generated/typed-codes/language-code.g.js";
import type { LocaleCode } from "../generated/typed-codes/locale-code.g.js";
import type { SubdivisionCode } from "../generated/typed-codes/subdivision-code.g.js";
import type { TimezoneCode } from "../generated/typed-codes/timezone-code.g.js";

/**
 * Mirror of .NET `D2.Shared.Geo.Abstractions.IGeoReference` —
 * strongly-typed lookup contract for the seven reference-data catalogs.
 * Every method takes a typed identifier (`*Code` real enum or wrapper
 * struct) and returns the single record shape — the type system enforces
 * the absence of a NotFound branch because the typed identifier IS the
 * catalog.
 *
 * Implementations live in `@d2/geo-default` backed by the codegen-emitted
 * static data; tests can supply ad-hoc fixtures by implementing this
 * interface directly.
 *
 * The lookup methods are exact-match by primary code — they never perform
 * name-based resolution. For free-form text → entity lookups use
 * `IGeoNameResolver` from `../name-resolution/i-geo-name-resolver.js`.
 *
 * Lookups return deprecated entities by default because deprecated codes
 * (`YU`, `SU`, etc.) must remain resolvable for hash citations + audit
 * replay. UI / dropdown code that wants to filter deprecated entries out
 * should walk the catalog arrays and check `entity.deprecation === undefined`
 * explicitly.
 */
export interface IGeoReference {
  /** Look up a `Country` by its typed code. */
  getCountry(code: CountryCode): Country;

  /** Look up a `Subdivision` by its branded ISO 3166-2 code. */
  getSubdivision(code: SubdivisionCode): Subdivision;

  /** Look up a `Timezone` by its branded IANA identifier. */
  getTimezone(code: TimezoneCode): Timezone;

  /** Look up a `Locale` by its branded IETF BCP-47 tag. */
  getLocale(code: LocaleCode): Locale;

  /** Look up a `Currency` by its typed code. */
  getCurrency(code: CurrencyCode): Currency;

  /** Look up a `Language` by its typed code. */
  getLanguage(code: LanguageCode): Language;

  /** Look up a `GeopoliticalEntity` by its typed code. */
  getGeopoliticalEntity(code: GeopoliticalEntityCode): GeopoliticalEntity;
}

File: public/packages/dotnet/geo/abstractions/IGeoReference.cs
// -----------------------------------------------------------------------
// <copyright file="IGeoReference.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Geo.Abstractions;

/// <summary>
/// Strongly-typed lookup contract for the seven reference-data catalogs.
/// Every method takes a typed identifier (real enum or wrapper struct)
/// and returns the matching non-nullable single-shape record — the type
/// system enforces the absence of a NotFound branch because the typed
/// identifier IS the catalog (every enum member maps to exactly one
/// entity; every wrapper struct value comes through closed-set validation
/// at the deserialization boundary).
/// </summary>
/// <remarks>
/// <para>
/// Inputs use the <c>*Code</c> suffix
/// (<see cref="CountryCode"/>, <see cref="CurrencyCode"/>,
/// <see cref="LanguageCode"/>, <see cref="GeopoliticalEntityCode"/>) for
/// closed-set enums and the wrapper struct names
/// (<see cref="SubdivisionCode"/>, <see cref="LocaleCode"/>,
/// <see cref="TimezoneCode"/>) for open-set codes.
/// </para>
/// <para>
/// String-input cousins (e.g. <c>TryGetCountry(string alpha2)</c>) and
/// free-form name resolution (<see cref="NameResolution.IGeoNameResolver"/>)
/// live outside this interface — they are boundary-code concerns and
/// MUST NOT be called from domain handlers.
/// </para>
/// </remarks>
public interface IGeoReference
{
    /// <summary>
    /// Resolves the full denormalized record for <paramref name="code"/>.
    /// </summary>
    /// <param name="code">The typed ISO 3166-1 alpha-2 identifier.</param>
    /// <returns>The country record (always non-null).</returns>
    Country GetCountry(CountryCode code);

    /// <summary>
    /// Resolves the full denormalized record for <paramref name="code"/>.
    /// </summary>
    /// <param name="code">The wrapped ISO 3166-2 subdivision code.</param>
    /// <returns>The subdivision record (always non-null).</returns>
    Subdivision GetSubdivision(SubdivisionCode code);

    /// <summary>
    /// Resolves the full denormalized record for <paramref name="code"/>.
    /// </summary>
    /// <param name="code">The wrapped IANA timezone identifier.</param>
    /// <returns>The timezone record (always non-null).</returns>
    Timezone GetTimezone(TimezoneCode code);

    /// <summary>
    /// Resolves the full denormalized record for <paramref name="code"/>.
    /// </summary>
    /// <param name="code">The wrapped IETF BCP 47 locale tag.</param>
    /// <returns>The locale record (always non-null).</returns>
    Locale GetLocale(LocaleCode code);

    /// <summary>
    /// Resolves the full denormalized record for <paramref name="code"/>.
    /// </summary>
    /// <param name="code">The typed ISO 4217 alpha identifier.</param>
    /// <returns>The currency record (always non-null).</returns>
    Currency GetCurrency(CurrencyCode code);

    /// <summary>
    /// Resolves the full denormalized record for <paramref name="code"/>.
    /// </summary>
    /// <param name="code">The typed ISO 639-1 identifier.</param>
    /// <returns>The language record (always non-null).</returns>
    Language GetLanguage(LanguageCode code);

    /// <summary>
    /// Resolves the full denormalized record for <paramref name="code"/>.
    /// </summary>
    /// <param name="code">The typed geopolitical-entity identifier.</param>
    /// <returns>The geopolitical-entity record (always non-null).</returns>
    GeopoliticalEntity GetGeopoliticalEntity(GeopoliticalEntityCode code);
}
